pkg/virtctl/expose/expose.go

```diff
@@ -0,0 +1,186 @@
+	cmd.Flags().StringVar(&ExternalIP, "external-ip", "", "Additional external IP address (not managed by the cluster) to accept for the service. If this IP is routed to a node, the service can be accessed by this IP in addition to its generated service IP. Optional.")
+	cmd.Flags().StringVar(&LoadBalancerIP, "load-balancer-ip", "", "IP to assign to the Load Balancer. If empty, an ephemeral IP will be created and used.")
+	cmd.Flags().Int32Var(&Port, "port", 0, "The port that the service should serve on")
+	cmd.MarkFlagRequired("port")
+	cmd.Flags().StringVar(&Protocol, "protocol", "TCP", "The network protocol for the service to be created.")
+	cmd.Flags().IntVar(&TargetPort, "target-port", 0, "Name or number for the port on the VM that the service should direct traffic to. Optional.")
+	cmd.Flags().Int32Var(&NodePort, "node-port", 0, "Port used to expose the service on each node in a cluster.")
+	cmd.Flags().StringVar(&ServiceType, "type", "ClusterIP", "Type for this service: ClusterIP, NodePort, or LoadBalancer.")
+	cmd.Flags().StringVar(&PortName, "port-name", "", "Name of the port. Optional.")
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+
+	return cmd
+}
+
+func usage() string {
+	usage := "  # Expose SSH to a virtual machine called 'myvm' as a node port (5555) of the cluster:\n"
+	usage += fmt.Sprintf("  virtctl expose myvm --port=5555 --target-port=22 --name=myvm-ssh --type=NodePort")
+	return usage
+}
+
+// executing the "expose" command
+func (o *Command) RunE(cmd *cobra.Command, args []string) error {
+	// first argument must be name of the VM
+	vmName := args[0]
+
+	// these are used to convert the flag values into service spec values
+	var protocol v1.Protocol
+	var targetPort intstr.IntOrString
+	var serviceType v1.ServiceType
+
+	// convert from integer to the IntOrString type
+	targetPort = intstr.FromInt(TargetPort)
+
+	// convert from string to the protocol enum
+	switch Protocol {
+	case "TCP":
+		protocol = v1.ProtocolTCP
+	case "UDP":
+		protocol = v1.ProtocolUDP
+	default:
+		return fmt.Errorf("Unknown protocol: %s", Protocol)
+	}
+
+	// convert from string to the service type enum
+	switch ServiceType {
+	case "ClusterIP":
+		serviceType = v1.ServiceTypeClusterIP
+	case "NodePort":
+		serviceType = v1.ServiceTypeNodePort
+	case "LoadBalancer":
+		serviceType = v1.ServiceTypeLoadBalancer
+	case "ExternalName":
+		return fmt.Errorf("Type: %s not supported", ServiceType)
+	default:
+		return fmt.Errorf("Unknown service type: %s", ServiceType)
+	}
+
+	// get the namespace
+	namespace, _, err := o.clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	// get the client
+	virtClient, err := kubecli.GetKubevirtClientFromClientConfig(o.clientConfig)
+	if err != nil {
+		return fmt.Errorf("Cannot obtain KubeVirt client: %v", err)
+	}
+
+	// TODO: not sure what are the options used for?
+	options := k8smetav1.GetOptions{}
+	// get the VM
```

Comment:
Instead of this fallback logic, what we will need, is another argument:

```
virtctl expose vm (vm) # should be created if the vm exists (no fallback to ovm in any way necessary)
virtctl expose ovm (ovm) # should be creaste if the ovm exists (labels can be obtained from spec.template.metadata.labels). No check if the vm exists or not is required.
```

So we need an extra argument which can have the values `vm,virtualmachine,vms,virtualmachines` or `ovm,ovms,offlinevirtualmachine,offlinevirtualmachines`. A third one would be `vmrs,vmrss,virtualmachinereplicasets,virtualmachinereplicaset`.

```
virtctl expose vmrs (vmrs) # should be creaste if the vm replica set exists (labels can be obtained from spec.template.metadata.labels). No check if vms exist are required.
```

Reply:
regarding replica sets:
1. your proposal is to use: ```vmrs.Spec.Template.ObjectMeta.Labels``` (when vmrs is the VM replica set object). Why not use: ```vmrs.Spec.Selector.MatchLabels``` ?
2. using replica set in: cluster/example/vm-replicaset-cirros.yaml there are 3 nodes being created.  To which ones of them does the cluster-ip belongs?